### DISCUSSION TITLE
Implementation of Stable PD controller: RNE, CRB and code migration

### DISCUSSION THREAD

**Discussion by rohit-kumar-j** [google-deepmind/mujoco#424] [Asking for Help] [answered]
I want to implement stable PD [Stable Proportional-Derivative Controllers](https://www.cc.gatech.edu/people/home/turk/my_papers/stable_pd.pdf) proposed here, which require the mass matrix and the forces needed to reach the given joint accelerations.

Reason to do this:Using larger time steps while keeping large gains stable throughout the simulation

Would It be right to assume that [`mjData.qfrc_inverse`](https://mujoco.readthedocs.io/en/latest/APIreference.html?highlight=qfrc_inverse#mjdata) is equivalent to `pybullet.calculateInverseDynamics()` 
and 
[`mujoco.mj_crb(model,data)`](https://mujoco.readthedocs.io/en/latest/APIreference.html?highlight=CRB#mj-crb) to `pybullet.calculateMassMatrix()`
(and accessed through [mjData.crb](https://mujoco.readthedocs.io/en/latest/APIreference.html?highlight=crb#mjdata)) in terms of implementing the Stable PD algorithm? Hence replacing them respectively in the following below would do the trick?

Simplified Stable PD from [the code here](https://github.com/bulletphysics/bullet3/blob/a1d96646c8ca28b99b2581dcfc4d74cc3b4de018/examples/pybullet/examples/pdControllerStable.py#L104) and [Leftmost in this deom video](https://www.youtube.com/watch?v=zwEEQd9nLz8&ab_channel=ErwinCoumans):

<details>
<Summary>Simplified Implementation</Summary>

```python
def computePD(self, bodyUniqueId, jointIndices, desiredPositions, desiredVelocities, kps, kds,
                maxForces, timeStep):
    numJoints = self._pb.getNumJoints(bodyUniqueId)
    jointStates = self._pb.getJointStates(bodyUniqueId, jointIndices)
    q1 = []
    qdot1 = []
    zeroAccelerations = []
    for i in range(numJoints):
      q1.append(jointStates[i][0])
      qdot1.append(jointStates[i][1])
      zeroAccelerations.append(0)

    q = np.array(q1)
    qdot = np.array(qdot1)
    qdes = np.array(desiredPositions)
    qdotdes = np.array(desiredVelocities)

    qError = qdes - q
    qdotError = qdotdes - qdot

    Kp = np.diagflat(kps)
    Kd = np.diagflat(kds)

    # Compute -Kp(q + qdot - qdes)
    p_term = Kp.dot(qError - qdot*timeStep)
    # Compute -Kd(qdot - qdotdes)
    d_term = Kd.dot(qdotError)

    # Compute Inertia matrix M(q)
    M = self._pb.calculateMassMatrix(bodyUniqueId, q1)
    M = np.array(M)
    # Given: M(q) * qddot + C(q, qdot) = T_ext + T_int
    # Compute Coriolis and External (Gravitational) terms G = C - T_ext
    G = self._pb.calculateInverseDynamics(bodyUniqueId, q1, qdot1, zeroAccelerations)
    G = np.array(G)
    # Obtain estimated generalized accelerations, considering Coriolis and Gravitational forces, and stable PD actions
    qddot = np.linalg.solve(a=(M + Kd * timeStep),
                            b=(-G + p_term + d_term))
    # Compute control generalized forces (T_int)
    tau = p_term + d_term - (Kd.dot(qddot) * timeStep)
    # Clip generalized forces to actuator limits
    maxF = np.array(maxForces)
    generalized_forces = np.clip(tau, -maxF, maxF)
    return generalized_forces
```
</details>

`pybullet.calculateInverseDynamics()`: < [Source: Pg-61,62](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#) >
> calculateInverseDynamics will compute the forces needed to reach the given joint accelerations, starting from specified joint positions and velocities. The inverse dynamics is computed using the Recursive Newton Euler algorithm (RNEA). calculateInverseDynamics returns a list of joint forces for each degree of freedom.

`pybullet.calculateMassMatrix()`: < [Source: Pg-63](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#)  >
> calculateMassMatrix will compute the system inertia for an articulated body given its joint positions. The composite rigid body algorithm (CBRA) is used to compute the mass matrix. The result is the square mass matrix with dimensions dofCount * dofCount, stored as a list of dofCount rows, each row is a list of dofCount mass matrix elements.

Centrifugal force, gravity, etc: (`C(q,q˙) - τext`, equation 9 from the paper) computed in the solver(mujoco doc reference below)< [Source](https://mujoco.readthedocs.io/en/latest/computation.html#general-framework) >
> The bias force cc includes Coriolis, centrifugal and gravitational forces. Their sum is computed using the Recursive Newton-Euler (RNE) algorithm with acceleration set to 0.

> The joint-space inertia matrix MM is computed using the Composite Rigid-Body (CRB) algorithm. This matrix is usually quite sparse, and we represent it as such, in a custom format tailored to kinematic trees.

**Comment by yuvaltassa** [comment, marked as answer]
If I understand correctly, you are looking for the bias forces and the mass matrix. These are computer during `mj_forward()` and then you can just read them out from `mjData`. You can also call sub-component functions if you want this to be faster. The bias forces live in `mjData.qfrc_bias` (computed by `mj_rne()`). The mass matrix lives in `mjData.qM` (computed by `mj_crb()`) but it is in sparse form. Call `mj_fullM()` to get the dense representation.

**Reply by Balint-H** [reply]
I've also been implementing stable PD but with C# for the Unity plugin. One difference from some bullet environments are the passive joint forces often used in Mujoco scenes I've seen before.

Is it correct to assume that the passive forces need to be added to the bias forces as well when calculating accelerations for SPD? (available in qfrc_passive)

**Reply by yuvaltassa** [reply]
I re-read the paper more thoroughly to see if I have an opinion, and it's now not clear to me why you need the mass matrix `M` and bias forces `C` in the first place. These are required for the physics, already computed by MuJoCo for the purpose of advancing the state, but do not come into the controller. 

The SPD equation is (10), which only depends on kinematic quantities. You can implement the first term directly using a [general actuator](https://mujoco.readthedocs.io/en/latest/computation.html#actuation-model). To get the second term which introduces an implicit dependence on $\ddot q$ and results in the implicit integration (11), switch your integrator to [`implicit`](https://mujoco.readthedocs.io/en/latest/computation.html#numerical-integration), this is [precisely what it does](https://github.com/deepmind/mujoco/blob/e96776c901e5a6f804e85e221bda8ecf85c26dbe/src/engine/engine_forward.c#L663).

Look at equation (11), instead of $M + \delta t K_d$ you can write $M - \delta t \frac{\partial \tau}{\partial \dot q}$. Compare to the implicit integrator documentation linked above.

Bottom line, as I understand it, these controllers can be implemented in MuJoCo as general actuators with the implicit integrator selected.

**Reply by rohit-kumar-j** [reply]
@yuvaltassa, can you give an example of this type of controller? I have tried to [make this here](https://github.com/rohit-kumar-j/test_spd_implementation). The StablePD implementation works well enough at lower (0.005s) and higher timesteps(0.05s), however, if I increase the DOFs of the system, by adding a body, there is shape conflict. 
Example:
`ValueError: operands could not be broadcast together with shapes (3,) (10,)` ([`in error_spd.py`](https://github.com/rohit-kumar-j/test_spd_implementation/blob/main/error_spd.py))

(The 7 added variables are perhaps, `x,y,z` for pos and `w,x,y,z` for orn of the newly added body?)

Here is a self-contained [`test implementation`](https://github.com/rohit-kumar-j/test_spd_implementation/blob/main/spd_utils.py#L5) (It uses `data.qpos` and `data.qvel`, fails at higher DOFs), is there a way to circumvent this?

**Reply by yuvaltassa** [reply]
Your issue is not higher dofs, it's quaternions. When there are quaternions in your model (free/ball joints) then `model.nq != model.nv`. The derivatives are of course in the nv-sized tangent space.

You need to use `mj_integratePos` and `mj_differentiatePos` (see [here](https://mujoco.readthedocs.io/en/latest/APIreference.html#mj-differentiatepos)) to go from the tangent space to the joint space and back.

Makes sense?
Let me know if you try and fail.

**Reply by rohit-kumar-j** [reply]
@yuvaltassa , even if we obtain the qpos values categorically, the issue is the mass matrix calculation, (or atleast that is what I think I am seeing here: [error_spd_utils.py](https://github.com/rohit-kumar-j/test_spd_implementation/blob/main/error_spd_utils.py), running this file:[error_spd.py](https://github.com/rohit-kumar-j/test_spd_implementation/blob/main/error_spd.py)

```python
    q = np.array(      # Not concerned with full qpos, as it contains [3 + 7((pos)x,y,z, (quat)w,x,y,z)] indices
        [
            data.qpos[0],
            data.qpos[1],
            data.qpos[2],
        ]
    )
    qdot = np.array(      # Not concerned with full qvel, as it contains [3 + 6(yaw pitch roll?)] indices
        [
            data.qvel[0],
            data.qvel[1],
            data.qvel[2],
        ]
    )
```
Error:
```bash
Kd: [[1900.    0.    0.]
 [   0. 1900.    0.]
 [   0.    0. 1900.]],np.shape(Kd):(3, 3)
Error at line line: a=(MassMatrix + Kd * model.opt.timestep),
ValueError: operands could not be broadcast together with shapes (9,9) (3,3)
```

**Reply by yuvaltassa** [reply]
Why are you using the full mass matrix? If you're taking only some dofs for the Kd callculation shouldn't you be doing the same for the mass matrix? I.e. I would expect `MassMatrix[:3, :3]`

**Reply by rohit-kumar-j** [reply]
How would we know which indices are for which robot/bodies? (Especially if it is mixed up in the xml file)
Is there a named accessor for the mass matrix?

**Reply by Balint-H** [reply]
The full mass matrix is organised in terms of DoF address, so if you store the subset of `qvel` indices of each joint (which you can get based on `jnt_dofadr`) in an array, then you can get the submatrix by indexing with that.

**Reply by rohit-kumar-j** [reply]
@Balint-H , are you referring to [this method](https://mujoco.readthedocs.io/en/latest/overview.html?highlight=jnt_dofadr#floating-objects)?
I have applied the `jnt_dofadr` like this:
```python
qposadr =model.jnt_qposadr[model.body_jntadr[mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_BODY, "my_floating_body")]]
qveladr =model.jnt_dofadr[model.body_jntadr[mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_BODY, "my_floating_body")]]
```
Is this how we should index the Mass Matrix and Bias Forces?:
```python
MassMatrix[:qveladr, :qveladr]
Bias_Forces[:qveladr]
```

 [my current code is here](https://github.com/rohit-kumar-j/test_spd_implementation/blob/main/error_spd_utils.py)

**Reply by Balint-H** [reply]
Not quite, you can store all dof indices in a single collection:

dof_indices = np.concatenate([get_all_joint_indices(joint_id) for joint in controlled_joint_ids])

Where get_all_joint_indices(joint_id) returns an iterable of all degrees of freedom addresses (associated with qv and qacc) associated with a joint (1 for hinge, 3 for ball). This method would use jnt_dofadr.

Then you can use fancy indexing with numpy to get the submatrix easily:

sub_M = full_M[dof_indices, dof_indices]

(no : needed)

Similarly with the bias forces.

sub_bias = bias[dof_indices]

**Reply by rohit-kumar-j** [reply]
Could you please elaborate on this?
> Where get_all_joint_indices(joint_id) returns an iterable of all degrees of freedom addresses (associated with qv and qacc) associated with a joint (1 for hinge, 3 for ball). This method would use jnt_dofadr.

**Reply by Balint-H** [reply]
It could be something like this:
(Please double check, didnt have time to test it, translated it from my current C# implementation). Also you might want to pass model as arguement depending on how this function is defined.

```python
def get_all_joint_indices(jnt_id):
    adr_start = model.jnt_dofadr[jnt_id]
    # assuming only ball, hinge or slide is used
    # Types of free, ball, slide, hinge: 0, 1, 2, 3
    n_dofs = 1 if model.jnt_type[jnt_id] > 1 else 3
    return np.arange(adr_start, adr_start+n_dofs)
```

**Reply by rohit-kumar-j** [reply]
I am getting errors with `get_all_joint_indices()`:

<details>
    <Summary>get_all_joint_indices</Summary>

```python
def get_all_joint_indices(model, jnt_name):
    jnt_id = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_JOINT, jnt_name)
    adr_start = model.jnt_dofadr[jnt_id]
    # assuming only ball, hinge or slide is used
    # Types of free, ball, slide, hinge: 0, 1, 2, 3
    if model.jnt_type[jnt_id] == 0:
        n_dofs = 6
        #  print("free")
    elif model.jnt_type[jnt_id] == 1:
        n_dofs = 3
        #  print("ball")
    elif model.jnt_type[jnt_id] == 2:
        n_dofs = 1
        #  print("slide")
    elif model.jnt_type[jnt_id] == 3:
        n_dofs = 1
        #  print("hinge")
    # n_dofs = 1 if model.jnt_type[jnt_id] > 1 else 3
    return np.arange(adr_start, adr_start + n_dofs)
```
</details>
The MM is referenced like this: MassMatrix[dof_indices, dof_indices] , however the error is:

```bash
a=(MassMatrix[dof_indices, dof_indices] + Kd * timeStep),
ValueError: operands could not be broadcast together with shapes (18,) (3,3)
```
[link_to_full_implementation](https://github.com/rohit-kumar-j/test_spd_implementation/blob/main/error_spd_utils.py)

Seems like I have too many errors. Could you kindly send me your C# implementation?

**Reply by saran-t** [reply]
@Balint-H @rohit-kumar-j If you're using the `mujoco` Python bindings, you should prefer the [named access API](https://mujoco.readthedocs.io/en/latest/python.html#named-access) over `mj_name2id`.

See e.g. https://colab.research.google.com/drive/1V5OdtiN1e2GlVd4RCp_KwSO6mZ0m5AGM#scrollTo=vXTUIbJy8kTr

**Reply by Balint-H** [reply]
@rohit-kumar-j 

I'm planning to put together the neccesary files for sharing but for that I first need to isolate the scripts so they are standalone and dont depend on other things in my project.

Until then, I found one bug in my suggestion now that I had a chance to look at it closer, MM should be indexed a different way to get the submatrix: `M[dof_indices, :][:, dof_indices]` (as numpy zips together multiple iterable indices in a single indexing operation, which we avoid with this)

Furthermore, I noticed you are using actuators to apply forces. If the joints are underactuated, then you should only store the dof indices of actuated dofs. Lets say there is `nu`<`nv` of them. So the dof_indices array should be (`nu`,) shaped. You should ensure the order of dof_indices matches the order of your collection of actuators.

Then indexing the mass matrix the above way should give you a `nu`x`nu` matrix.

Kd should be the same shape as well.

**Reply by yuvaltassa** [reply]
I think @saran-t meant to link to the relevant [docs section](https://mujoco.readthedocs.io/en/latest/python.html#named-access). Also note that you get infomative messages if you do (in the above colab) `m.joint('abdomen_x')` or `m.joint()`

**Reply by saran-t** [reply]
@yuvaltassa is right, I'll correct the link in my previous comment for future readers :)

**Reply by rohit-kumar-j** [reply]
@Balint-H , thanks for the explanation!
I was able to get a stable simulation at `timestep=0.01s`: (Will upload the code once I clean it up)

https://user-images.githubusercontent.com/37873142/190503669-2d8c33ab-b07d-4ae7-9a47-5d6e214adeb5.mp4

 @yuvaltassa , @saran-t , thanks for the support!
